spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/LocalOptimizer.scala

```diff
@@ -152,8 +152,31 @@ class LocalOptimizer[T: ClassTag] (
           })
       )
       val loss = lossSum / parallelism
-      grad.div(ev.fromType(parallelism))
+      var scale = ev.fromType(parallelism)
+      if (gradientClippingParams.enableL2NormClipping) {
+        val squares = new Array[Double](syncGradParallelNum)
+        Engine.default.invokeAndWait((0 until syncGradParallelNum).map(tid => () => {
+          val offset = tid * syncGradTaskSize + math.min(tid, syncGradExtraTask)
+          val length = syncGradTaskSize + (if (tid < syncGradExtraTask) 1 else 0)
+          squares(tid) = ev.toType[Double](grad.narrow(1, offset + 1, length).sumSquare())
+        }))
+        var sum = 0.0
+        var i = 0
+        while (i < squares.size) {
+          sum += squares(i)
+          i += 1
+        }
+        val l2Norm = (math.sqrt(sum) / parallelism).toFloat
+        
+        if (l2Norm > gradientClippingParams.normValueClip) {
+          scale = ev.fromType[Float]((l2Norm * parallelism) / gradientClippingParams.normValueClip)
+        }
+      }
+      grad.div(scale)
 
+      if (gradientClippingParams.enableConstantClipping) {
```

Comment:
possible to make it a trait? so that people could define their own clipping.

Reply:
I think we can leave it for future improvement.